lang/py/avro/units.py

```diff
@@ -0,0 +1,418 @@
+        # for example an error or warning if corresponding fields do not
+        # both have unit annotations.
+        if "unit" not in rf.props: continue
+        # Fields that were not in the write schema do not require resolving.
+        if rfname not in writer_schema.fields_dict: continue
+        wf = writer_schema.fields_dict[rfname]
+        if "unit" not in wf.props: continue
+        # Unit resolutions are only defined for numeric field types.
+        if (rf.props['type']).type not in {'float', 'double', 'int', 'long'}:
+          # possible policies: warning or error here
+          continue
+        # if units are compatible, then we'll get a coefficient of conversion:
+        ruexpr = _parse_unit_expr(rf.props["unit"])
+        wuexpr = _parse_unit_expr(wf.props["unit"])
+        coef = self.coefficient(wuexpr, ruexpr)
+        if coef is None:
+          # throwing an exception here might make returning true/false irrelevant
+          raise UnitException('Field %s failed unit resolution during schema matching' % (rfname))
+        rf.props[wsid] = coef
+      # tag this write schema so we don't need to resolve units in the future
+      setattr(reader_schema, wsid, True)
+      return True
+    return False
+
+class CanonicalUnit(object):
+  def __init__(self, coef, cumap):
+    if not isinstance(coef, UnitCoef):
+      raise UnitException('coef was not a UnitCoef: "%s"' % (coef))
+    if not isinstance(cumap, dict):
+      raise UnitException('cumap was not a dict: "%s"' % (cumap))
+    self._coef = coef
+    self._map = cumap
+
+  coef = property(lambda self: self._coef)
+  map = property(lambda self: self._map)
+
+  def mul(self, rhs):
+    lmap = copy.deepcopy(self.map)
+    rmap = rhs.map
+    for (u, re) in rmap.items():
+      e = (lmap.get(u) or 0) + re
+      if e == 0:
+        if u in lmap: del lmap[u]
+      else:
+        lmap[u] = e
+    return CanonicalUnit(self.coef.mul(rhs.coef), lmap)
+
+  def div(self, rhs):
+    lmap = copy.deepcopy(self.map)
+    rmap = rhs.map
+    for (u, re) in rmap.items():
+      e = (lmap.get(u) or 0) - re
+      if e == 0:
+        if u in lmap: del lmap[u]
+      else:
+        lmap[u] = e
+    return CanonicalUnit(self.coef.div(rhs.coef), lmap)
+
+  def pow(self, exp):
+    if exp == 0:
+      return CanonicalUnit(1.0, {})
+    lmap = copy.deepcopy(self.map)
+    if exp == 1:
+      return CanonicalUnit(self.coef, lmap)
+    for u in lmap.keys():
+      lmap[u] = exp * lmap[u]
+    return CanonicalUnit(self.coef.pow(exp), lmap)
+
+  def __str__(self):
+    return "CanonicalUnit(%s, %s)" % (self.coef, self.map)
+
+  def __repr__(self):
+    return str(self)
+
+class UnitDef(object):
+  def __init__(self, prop, name, abbv):
+    self._prop = prop
+    self._name = name
+    self._abbv = abbv
+
+  prop = property(lambda self: self._prop)
+  name = property(lambda self: self._name)
+  abbv = property(lambda self: self._abbv)
+
+  def __repr__(self):
+    return str(self)
```

Comment:
the built-in repr wasn't giving me a human readable string, IIRC.  either that or I was doing something wrong.

Reply:
That's interesting. I'd suggest steering clear of making `repr` and `str` the same. Or more's the point, please be very careful not to let the `repr` look like any other type of object. I think there was a case above where a Units type object had a `repr` that made it look exactly like a dictionary. That's definitely a situation to avoid.